009-ui-and-admin-server-security.md

```diff
@@ -0,0 +1,84 @@
+
+![Strimzi UI using SCRAM-SHA512](./images/scram-sha512-ui.png)
+
+This Kafka client model translates to a UI and Admin Server in a similar way.
+When the user loads up the UI in a browser, they are presented with a dialogue box requiring them to enter a userId and password.
+The userId and password they enter are pre-defined to Kafka by the creation of a Kafka user through the `User Operator` or the Kafka `bin/kafka-configs.sh ` command line tool.
+Once the userId and password have been entered, the UI validates the credentials on Zookeeper using `passport.js`.
+The result that is returned will be marked as valid or invalid.
+If the credentials are valid, they are stored in the session data associated with that browser session.
+When the UI requires data from the Admin Server, the credentials are recovered from the session data and placed as an auth header on the request.
+The Admin Server receives the request and configures the Kafka client in preparation for calling the Kafka broker.
+The SCRAM credentials are retrieved from the auth header and are added to the Kafka client configuration properties before executing the call to the Kafka broker.
+From thereon, the Kafka authentication and authorization mechanism works in an identical way to the description above for the Kafka Clients.
+
+This mechanism should be straightforward to setup using a `passport.js` strategy.
+The Strimzi UI is served from an Express server and that Express server will be inside the Strimzi cluster allowing the Express Server to talk to the Zookeeper on the secure port 2181.
+The login would not go through the Admin Server which hides the process from the external network but it may be desirable to add a `login` endpoint to the Admin Server which would allow session based applications to use the Admin Server in a natural way.
+
+The Strimzi Admin Server API could be called directly by the user creating an auth header on the request.
+Like the UI flow, the credentials would be retrieved from the auth header and placed into the Kafka client properties as SCRAM credentials.
+
+## OAuth 2
+
+![Kafka Client using Oauth2](./images/oauth-client.png)
+
+In the OAuth 2 scenario, Strimzi currently requires the user to configure the Kafka client with a clientId and secret.
+The client will then obtain an access token from the identity provider which is KeyCloak and that token flows to the broker.
+In the broker, calls to KeyCloak will be made based on the type of token that was requested by the client and passed to the broker.
+
+![Strimzi UI using Oauth2](./images/oauth-ui.png)
+
+In the UI model, the user will be presented with a login dialogue box when the UI is loaded into the browser and requires the user to enter a userId and password.
+The userId needs to be pre-defined to KeyCloak through the KeyCloak user interface.
+When the user enters their userId and password, the UI will contact KeyCloak to authorise the user and obtain an access token.
```

Comment:
How is 'UI will contact Keycloak' to be understood? There's UI as 'the server side app with REST API' and the UI as 'the browser JS app'. Authorization servers like Keycloak are used as Single Sign-On servers where user's browser is completely redirected to a central login page, and after successful login redirected back to web app, so that the applications never see the password, not even the browser JS app - that's the original reason OAuth was developed. None of the applications ever receives the user's password, they only receive the access token, which grants these apps limited access for a limited time, after which the apps redirect user to SSO server to login again. In practice the redirect is often invisible because there is a mechanism to issue a new access token without user interaction in certain circumstances (when using a refresh token).